ipc/src/platform/platform_handle.rs

```diff
@@ -0,0 +1,200 @@
+    pub unsafe fn to_any_type<U>(self) -> PlatformHandle<U> {
+        PlatformHandle {
+            fd: self.fd,
+            inner: self.inner,
+            phantom: PhantomData,
+        }
+    }
+
+    /// OwnedFileHandle innertype is safe to instantiate via into_instance
+    pub fn to_untyped(self) -> PlatformHandle<OwnedFileHandle> {
+        unsafe { self.to_any_type() }
+    }
+}
+
+impl<T> TransferHandles for PlatformHandle<T> {
+    fn move_handles<Transport: crate::handles::HandlesTransport>(
+        &self,
+        transport: Transport,
+    ) -> Result<(), Transport::Error> {
+        transport.move_handle(self.clone())
+    }
+
+    fn receive_handles<Transport: crate::handles::HandlesTransport>(
+        &mut self,
+        transport: Transport,
+    ) -> Result<(), Transport::Error> {
+        let received_handle = transport.provide_handle(self)?;
+        self.inner = received_handle.inner;
+        Ok(())
+    }
+}
+
+impl<T> PlatformHandle<T>
+where
+    T: FilelikeViewType,
+{
+    pub fn as_filelike_view(&self) -> io::Result<FilelikeView<'_, T>> {
+        Ok(self.as_owned_fd()?.as_filelike_view())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{fs::File, io::Write, thread};
+
+    use super::PlatformHandle;
+    macro_rules! assert_file_is_open_for_writing {
+        ($file:expr) => {{
+            writeln!($file, "test").unwrap();
+        }};
+    }
+
+    #[test]
+    #[cfg_attr(miri, ignore)]
+    fn test_platform_handles_fd_handling() {
+        let mut file = tempfile::tempfile().unwrap();
+        assert_file_is_open_for_writing!(file);
+
+        let shared = PlatformHandle::from(file);
+
+        {
+            let clone = shared.clone();
+            // can't uniquely own a shared instance with multiple owners
+            shared.clone().into_owned_handle().unwrap_err();
+            clone.into_owned_handle().unwrap_err();
+        }
+        // once no one is using the instance we can convert it to owned
+        let owned = shared.into_owned_handle().unwrap();
+        let mut file: File = owned.into();
+
+        assert_file_is_open_for_writing!(file);
+    }
+
+    #[test]
+    #[cfg_attr(miri, ignore)]
+    fn test_platform_handle_fd_borrowing() {
+        let mut file = tempfile::tempfile().unwrap();
+        assert_file_is_open_for_writing!(file);
+
+        let shared = PlatformHandle::from(file);
+        let mut joins = vec![];
+        for _ in 0..100 {
+            let shared = shared.clone();
+            let th = thread::spawn(move || {
+                let mut file = &*shared.as_filelike_view().unwrap();
+                assert_file_is_open_for_writing!(file);
+                1
+            });
+            joins.push(th);
+        }
+        let cnt: i32 = joins.into_iter().map(|j| j.join().unwrap()).sum();
+        assert_eq!(100, cnt);
+
+        let mut file = &*shared.as_filelike_view().unwrap();
+        assert_file_is_open_for_writing!(file);
```

Comment:
I'm not sure if this is really testing that the ref counting works in the end.
Shouldn't we turn `shared` into an `OwnedFd` like this?
```rust
let owned = shared.into_owned_handle().unwrap();
let mut file: File = owned.into();
assert_file_is_open_for_writing!(file);
```

Reply:
The previous test is already testing that, basically, I think?
Not sure, I haven't written that code, it just got moved over from unix/ dir.